File: 6010.完成旅途的最少时间/s.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

int main() { return 0; }

class Solution {
public:
    long long minimumTime(vector<int>& time, int ts) {
        using LL = long long;

        LL lo = 0, hi = (LL)ts * (*min_element(time.begin(), time.end())) + 1;

        while (lo < hi) {
            LL mi = (hi - lo) / 2 + lo;
            LL k = 0;
            for (auto x : time) {
                k += mi / x;
            }
            if (k < ts) {
                lo = mi + 1;
            } else {
                hi = mi;
            }
        }

        return lo;
    }
};

File: 6010.完成旅途的最少时间/s.go
package main

import "math"

func minimumTime(time []int, totalTrips int) int64 {
    lo := int64(0)
    _, vMin := MinElement(time)
    hi := int64(totalTrips) * int64(vMin)

    for lo < hi {
        mi := (hi - lo) / 2 + lo
        k := int64(0)
        for _, t := range time {
            k += mi / int64(t)
        }
        if k < int64(totalTrips) {
            lo = mi + 1
        } else {
            hi = mi
        }
    }

    return lo
}

func MinElement(arr []int) (idx, val int) {
    idx = -1
    val = math.MaxInt

    for i, v := range arr {
        if v < val {
            idx = i
            val = v
        }
    }
    return
}